### PR TITLE
Merge/remove some GPIO-related macros

### PR DESCRIPTION
**Pull request by bugadani**
Stripped out of #2341, focusing mainly on reducing the number of necessary macro calls in target-specific code.

### CODE REVIEW

File: esp-hal/src/gpio/mod.rs [hunk 1]
```diff
@@ -987,6 +1007,27 @@ macro_rules! io_type {
     (Output, $gpionum:literal) => {
         impl $crate::gpio::OutputPin for GpioPin<$gpionum> {}
     };
+    (Analog, $gpionum:literal) => {
+        // FIXME: the implementation shouldn't be in the GPIO module
```
Comment (Dominaezzz) [inline esp-hal/src/gpio/mod.rs:1011]: Where else would you put this?

File: esp-hal/src/gpio/mod.rs [hunk 2]
```diff
@@ -987,6 +1007,27 @@ macro_rules! io_type {
...
+        #[cfg(any(esp32c2, esp32c3, esp32c6, esp32h2))]
+        impl $crate::gpio::AnalogPin for GpioPin<$gpionum> {
+            /// Configures the pin for analog mode.
+            fn set_analog(&self, _: $crate::private::Internal) {
```
Comment (Dominaezzz) [inline esp-hal/src/gpio/mod.rs:1015]: I hope to see this method's implementation can be moved into the `AnalogPin` trait itself in future.

**Review by Dominaezzz — APPROVED** [review]
Can't say I fully understand the macro magic here but if the HIL tests are still building and passing, it should be working just fine

File: esp-hal/src/gpio/mod.rs (hunk 1 above)
Comment (bugadani) [reply esp-hal/src/gpio/mod.rs:1011]: The implementation itself is target-specific, it has no place in gpio/mod.rs. I don't yet know how to organize it, though, but I want to keep this macro just a bunch of "if this, then impl this trait in one line" cases, like Input and Output are right now.

**Pull request merged**